File: src/data/generateData.js
// Synthetic Healthcare Claims Data Generator
// All data is entirely fictional for demonstration purposes

const REGIONS = ['Northeast', 'Southeast', 'Midwest', 'Southwest', 'West'];

const DIAGNOSES = [
  { code: 'I21', name: 'Acute MI', category: 'Cardiovascular' },
  { code: 'J18', name: 'Pneumonia', category: 'Respiratory' },
  { code: 'E11', name: 'Type 2 Diabetes', category: 'Endocrine' },
  { code: 'M54', name: 'Back Pain', category: 'Musculoskeletal' },
  { code: 'F32', name: 'Depression', category: 'Mental Health' },
  { code: 'N18', name: 'Chronic Kidney Disease', category: 'Renal' },
  { code: 'I10', name: 'Hypertension', category: 'Cardiovascular' },
  { code: 'J44', name: 'COPD', category: 'Respiratory' },
  { code: 'K92', name: 'GI Hemorrhage', category: 'Gastrointestinal' },
  { code: 'A41', name: 'Sepsis', category: 'Infectious Disease' },
];

const PROCEDURES = [
  { code: '99213', name: 'Office Visit', baseCost: 180 },
  { code: '99223', name: 'Inpatient Admission', baseCost: 3200 },
  { code: '93000', name: 'ECG', baseCost: 220 },
  { code: '71046', name: 'Chest X-Ray', baseCost: 340 },
  { code: '80053', name: 'Metabolic Panel', baseCost: 110 },
  { code: '43239', name: 'GI Endoscopy', baseCost: 1800 },
  { code: '27447', name: 'Knee Replacement', baseCost: 28000 },
  { code: '33533', name: 'CABG', baseCost: 65000 },
  { code: '70553', name: 'MRI Brain', baseCost: 2400 },
  { code: '99291', name: 'Critical Care', baseCost: 900 },
];

const PAYERS = ['Medicare', 'Medicaid', 'Commercial', 'Self-Pay', 'CHIP'];
const FACILITY_TYPES = ['Hospital', 'Outpatient Clinic', 'Urgent Care', 'Specialty Center', 'Rehab Facility'];

function seededRand(seed) {
  let s = seed;
  return function () {
    s = (s * 1664525 + 1013904223) & 0xffffffff;
    return (s >>> 0) / 4294967295;
  };
}

export function generateClaims(count = 2000) {
  const rand = seededRand(42);
  const claims = [];

  for (let i = 0; i < count; i++) {
    const r = () => rand();
    const region = REGIONS[Math.floor(r() * REGIONS.length)];
    const diagnosis = DIAGNOSES[Math.floor(r() * DIAGNOSES.length)];
    const procedure = PROCEDURES[Math.floor(r() * PROCEDURES.length)];
    const payer = PAYERS[Math.floor(r() * PAYERS.length)];
    const facilityType = FACILITY_TYPES[Math.floor(r() * FACILITY_TYPES.length)];

    const year = 2022 + Math.floor(r() * 3);
    const month = Math.floor(r() * 12);
    const day = Math.floor(r() * 28) + 1;
    const serviceDate = new Date(year, month, day);

    const age = Math.floor(r() * 65) + 18;
    const gender = r() > 0.48 ? 'Female' : 'Male';

    const costMultiplier =
      0.6 + r() * 0.9 + (region === 'Northeast' || region === 'West' ? 0.3 : 0);
    const billedAmount = Math.round(procedure.baseCost * costMultiplier * 100) / 100;
    const allowedAmount = Math.round(billedAmount * (0.55 + r() * 0.3) * 100) / 100;
    const paidAmount = Math.round(
      allowedAmount * (payer === 'Self-Pay' ? 0.3 + r() * 0.2 : 0.8 + r() * 0.15) * 100
    ) / 100;
    const patientResponsibility = Math.round((allowedAmount - paidAmount) * 100) / 100;

    const losBase =
      diagnosis.category === 'Cardiovascular' ? 5
      : diagnosis.category === 'Respiratory' ? 4
      : 2;
    const lengthOfStay =
      facilityType === 'Hospital' ? Math.floor(r() * losBase * 2) + 1 : 0;

    const readmitProb =
      (diagnosis.category === 'Cardiovascular' ? 0.18 : 0.08) +
      (age > 65 ? 0.08 : 0) +
      (payer === 'Medicaid' ? 0.05 : 0);
    const readmitted = r() < readmitProb;
    const qualityScore = Math.round((0.5 + r() * 0.5) * 100) / 100;
    const preventable = r() < 0.12;

    claims.push({
      claimId: `CLM${String(i + 1).padStart(6, '0')}`,
      patientId: `PT${String(Math.floor(r() * 800) + 1).padStart(5, '0')}`,
      serviceDate: serviceDate.toISOString().split('T')[0],
      year,
      month: month + 1,
      region,
      facilityType,
      diagnosis,
      procedure,
      payer,
      age,
      ageGroup:
        age < 35 ? '18-34'
        : age < 50 ? '35-49'
        : age < 65 ? '50-64'
        : '65+',
      gender,
      billedAmount,
      allowedAmount,
      paidAmount,
      patientResponsibility,
      lengthOfStay,
      readmitted,
      qualityScore,
      preventable,
    });
  }
  return claims;
}

export function computeKPIs(claims) {
  const total = claims.length;
  const totalBilled = claims.reduce((s, c) => s + c.billedAmount, 0);
  const totalPaid = claims.reduce((s, c) => s + c.paidAmount, 0);
  const readmissions = claims.filter((c) => c.readmitted).length;
  const hospitalClaims = claims.filter((c) => c.lengthOfStay > 0);
  const avgLOS =
    hospitalClaims.length > 0
      ? hospitalClaims.reduce((s, c) => s + c.lengthOfStay, 0) / hospitalClaims.length
      : 0;
  const preventable = claims.filter((c) => c.preventable).length;
  const avgQuality = claims.reduce((s, c) => s + c.qualityScore, 0) / total;
  return {
    totalClaims: total,
    totalBilled: Math.round(totalBilled),
    totalPaid: Math.round(totalPaid),
    avgCostPerClaim: Math.round(totalPaid / total),
    readmissionRate: Math.round((readmissions / total) * 1000) / 10,
    avgLengthOfStay: Math.round(avgLOS * 10) / 10,
    preventableClaims: preventable,
    preventableRate: Math.round((preventable / total) * 1000) / 10,
    avgQualityScore: Math.round(avgQuality * 1000) / 10,
  };
}

export function costByRegion(claims) {
  const map = {};
  claims.forEach((c) => {
    if (!map[c.region])
      map[c.region] = { region: c.region, totalPaid: 0, count: 0, readmissions: 0 };
    map[c.region].totalPaid += c.paidAmount;
    map[c.region].count++;
    if (c.readmitted) map[c.region].readmissions++;
  });
  return Object.values(map).map((r) => ({
    ...r,
    avgCost: Math.round(r.totalPaid / r.count),
    readmissionRate: Math.round((r.readmissions / r.count) * 1000) / 10,
  }));
}

export function utilizationByDiagnosis(claims) {
  const map = {};
  claims.forEach((c) => {
    const key = c.diagnosis.name;
    if (!map[key])
      map[key] = { name: key, category: c.diagnosis.category, count: 0, totalCost: 0, readmissions: 0 };
    map[key].count++;
    map[key].totalCost += c.paidAmount;
    if (c.readmitted) map[key].readmissions++;
  });
  return Object.values(map)
    .map((d) => ({
      ...d,
      avgCost: Math.round(d.totalCost / d.count),
      readmissionRate: Math.round((d.readmissions / d.count) * 1000) / 10,
    }))
    .sort((a, b) => b.count - a.count);
}

export function trendByMonth(claims) {
  const map = {};
  claims.forEach((c) => {
    const key = `${c.year}-${String(c.month).padStart(2, '0')}`;
    if (!map[key])
      map[key] = { period: key, year: c.year, month: c.month, count: 0, totalCost: 0, readmissions: 0 };
    map[key].count++;
    map[key].totalCost += c.paidAmount;
    if (c.readmitted) map[key].readmissions++;
  });
  return Object.values(map)
    .sort((a, b) => a.period.localeCompare(b.period))
    .map((m) => ({
      ...m,
      avgCost: Math.round(m.totalCost / m.count),
      readmissionRate: Math.round((m.readmissions / m.count) * 1000) / 10,
    }));
}

export function costByPayer(claims) {
  const map = {};
  claims.forEach((c) => {
    if (!map[c.payer])
      map[c.payer] = { payer: c.payer, count: 0, totalPaid: 0, totalBilled: 0 };
    map[c.payer].count++;
    map[c.payer].totalPaid += c.paidAmount;
    map[c.payer].totalBilled += c.billedAmount;
  });
  return Object.values(map).map((p) => ({
    ...p,
    avgCost: Math.round(p.totalPaid / p.count),
    paymentRatio: Math.round((p.totalPaid / p.totalBilled) * 1000) / 10,
  }));
}

export function qualityMetrics(claims) {
  const map = {};
  claims.forEach((c) => {
    const cat = c.diagnosis.category;
    if (!map[cat])
      map[cat] = { category: cat, count: 0, totalQuality: 0, preventable: 0, readmissions: 0 };
    map[cat].count++;
    map[cat].totalQuality += c.qualityScore;
    if (c.preventable) map[cat].preventable++;
    if (c.readmitted) map[cat].readmissions++;
  });
  return Object.values(map).map((c) => ({
    ...c,
    avgQuality: Math.round((c.totalQuality / c.count) * 1000) / 10,
    preventableRate: Math.round((c.preventable / c.count) * 1000) / 10,
    readmissionRate: Math.round((c.readmissions / c.count) * 1000) / 10,
  }));
}

File: python/generate_data.py
"""
Healthcare Claims Data Generator
Generates synthetic patient-level healthcare claims datasets
"""

import pandas as pd
import numpy as np
import random
from datetime import datetime, timedelta
import os

np.random.seed(42)
random.seed(42)

# ── Constants ──────────────────────────────────────────────────────────────────
N_PATIENTS    = 2_000
N_CLAIMS      = 15_000
REGIONS       = ["Northeast", "Southeast", "Midwest", "West", "Southwest"]
DIAGNOSES     = {
    "E11":  "Type 2 Diabetes",
    "I10":  "Hypertension",
    "J45":  "Asthma",
    "M54":  "Back Pain",
    "F32":  "Depression",
    "I25":  "Coronary Artery Disease",
    "N18":  "Chronic Kidney Disease",
    "J44":  "COPD",
    "Z51":  "Chemotherapy",
    "S72":  "Hip Fracture",
}
PROCEDURES    = {
    "99213": "Office Visit",
    "93000": "EKG",
    "71046": "Chest X-Ray",
    "80053": "Comprehensive Metabolic Panel",
    "85025": "Complete Blood Count",
    "99232": "Subsequent Hospital Care",
    "27447": "Knee Replacement",
    "43239": "Upper GI Endoscopy",
    "99285": "Emergency Dept Visit",
    "36415": "Blood Draw",
}
PAYERS        = ["Medicare", "Medicaid", "BlueCross", "Aetna", "UnitedHealth", "Self-Pay"]
FACILITY_TYPES = ["Hospital", "Outpatient Clinic", "ER", "Specialist Office", "Primary Care"]
GENDERS       = ["M", "F", "Other"]

START_DATE = datetime(2022, 1, 1)
END_DATE   = datetime(2024, 12, 31)


def random_date(start=START_DATE, end=END_DATE):
    delta = end - start
    return start + timedelta(days=random.randint(0, delta.days))


# ── 1. Patients ────────────────────────────────────────────────────────────────
def generate_patients():
    print("Generating patients …")
    patients = []
    for pid in range(1, N_PATIENTS + 1):
        age    = int(np.random.normal(55, 18))
        age    = max(18, min(95, age))
        region = random.choice(REGIONS)
        # region-based cost multiplier stored for later use
        patients.append({
            "patient_id":    pid,
            "age":           age,
            "gender":        random.choice(GENDERS),
            "region":        region,
            "payer":         random.choice(PAYERS),
            "zip_code":      f"{random.randint(10000, 99999)}",
            "chronic_conditions": random.randint(0, 5),
            "enrollment_date": random_date(START_DATE, START_DATE + timedelta(days=365)).strftime("%Y-%m-%d"),
        })
    return pd.DataFrame(patients)


# ── 2. Claims ──────────────────────────────────────────────────────────────────
def generate_claims(patients_df):
    print("Generating claims …")
    region_multiplier = {
        "Northeast": 1.25, "West": 1.20,
        "Midwest":   1.00, "Southeast": 0.92, "Southwest": 0.88,
    }
    claims = []
    diag_codes = list(DIAGNOSES.keys())
    proc_codes = list(PROCEDURES.keys())

    for cid in range(1, N_CLAIMS + 1):
        patient  = patients_df.sample(1).iloc[0]
        pid      = patient["patient_id"]
        region   = patient["region"]
        payer    = patient["payer"]
        age      = patient["age"]
        chronic  = patient["chronic_conditions"]

        diag  = random.choice(diag_codes)
        proc  = random.choice(proc_codes)
        ftype = random.choice(FACILITY_TYPES)

        # Base cost varies by facility and complexity
        base_cost = {
            "Hospital": np.random.lognormal(8.5, 0.8),
            "ER":       np.random.lognormal(7.8, 0.7),
            "Outpatient Clinic": np.random.lognormal(6.5, 0.6),
            "Specialist Office": np.random.lognormal(6.0, 0.5),
            "Primary Care":      np.random.lognormal(5.5, 0.4),
        }[ftype]

        # Adjust for age, chronic conditions, region
        cost = base_cost * (1 + age / 200) * (1 + chronic * 0.1) * region_multiplier[region]
        cost = round(cost, 2)

        # Payer adjustments
        paid_pct = {"Medicare": 0.80, "Medicaid": 0.65, "BlueCross": 0.78,
                    "Aetna": 0.75, "UnitedHealth": 0.77, "Self-Pay": 0.45}[payer]
        paid     = round(cost * paid_pct, 2)

        svc_date = random_date()
        los      = random.randint(0, 14) if ftype == "Hospital" else 0  # length of stay
        readmit  = 1 if (ftype == "Hospital" and random.random() < (0.08 + chronic * 0.02)) else 0

        claims.append({
            "claim_id":        cid,
            "patient_id":      pid,
            "service_date":    svc_date.strftime("%Y-%m-%d"),
            "service_year":    svc_date.year,
            "service_month":   svc_date.month,
            "service_quarter": f"Q{(svc_date.month - 1)//3 + 1}",
            "diagnosis_code":  diag,
            "diagnosis_desc":  DIAGNOSES[diag],
            "procedure_code":  proc,
            "procedure_desc":  PROCEDURES[proc],
            "facility_type":   ftype,
            "region":          region,
            "payer":           payer,
            "total_cost":      cost,
            "paid_amount":     paid,
            "patient_oop":     round(cost - paid, 2),
            "length_of_stay":  los,
            "readmission_30d": readmit,
            "claim_status":    random.choice(["Approved", "Approved", "Approved", "Denied", "Pending"]),
        })

    return pd.DataFrame(claims)


# ── 3. Quality Metrics ─────────────────────────────────────────────────────────
def generate_quality_metrics(claims_df):
    print("Generating quality metrics …")
    metrics = []
    for region in REGIONS:
        for year in [2022, 2023, 2024]:
            sub = claims_df[(claims_df["region"] == region) & (claims_df["service_year"] == year)]
            if sub.empty:
                continue
            hosp = sub[sub["facility_type"] == "Hospital"]
            metrics.append({
                "region":               region,
                "year":                 year,
                "total_claims":         len(sub),
                "avg_cost_per_claim":   round(sub["total_cost"].mean(), 2),
                "total_spend":          round(sub["total_cost"].sum(), 2),
                "readmission_rate_pct": round(hosp["readmission_30d"].mean() * 100, 2) if len(hosp) else 0,
                "denial_rate_pct":      round((sub["claim_status"] == "Denied").mean() * 100, 2),
                "avg_los_hospital":     round(hosp["length_of_stay"].mean(), 2) if len(hosp) else 0,
                "er_visit_rate_pct":    round((sub["facility_type"] == "ER").mean() * 100, 2),
            })
    return pd.DataFrame(metrics)


# ── Main ───────────────────────────────────────────────────────────────────────
def main():
    out = os.path.join(os.path.dirname(__file__), "..", "data")
    os.makedirs(out, exist_ok=True)

    patients = generate_patients()
    claims   = generate_claims(patients)
    quality  = generate_quality_metrics(claims)

    patients.to_csv(f"{out}/patients.csv",        index=False)
    claims.to_csv(  f"{out}/claims.csv",           index=False)
    quality.to_csv( f"{out}/quality_metrics.csv",  index=False)

    print(f"\n✅  Datasets saved to {os.path.abspath(out)}/")
    print(f"   patients.csv        → {len(patients):,} rows")
    print(f"   claims.csv          → {len(claims):,}  rows")
    print(f"   quality_metrics.csv → {len(quality):,}    rows")
    return patients, claims, quality


if __name__ == "__main__":
    main()
